File: scaffold/container-setup.js
#!/usr/bin/env node
/**
 * container-setup.js — Runs INSIDE the Docker container at startup.
 * Sets up the Chrome browser bridge and configures Claude Code settings.
 * Claude Code itself is installed at build time in the Dockerfile.
 */

import { execFileSync, spawn } from 'node:child_process';
import { existsSync, readFileSync, mkdirSync, writeFileSync, chmodSync, openSync, closeSync } from 'node:fs';
import { join } from 'node:path';

function log(msg) {
  const ts = new Date().toTimeString().slice(0, 8);
  console.log(`[setup] ${ts} ${msg}`);
}

function which(cmd) {
  try {
    execFileSync('which', [cmd], { stdio: 'ignore' });
    return true;
  } catch { return false; }
}

// -- Chrome bridge setup -----------------------------------------------------

function setupChromeBridge() {
  const bridgePort = process.env.BRIDGE_PORT || '9229';
  const user = process.env.USER;
  const home = process.env.HOME;

  const sockDir = `/tmp/claude-mcp-browser-bridge-${user}`;
  mkdirSync(sockDir, { recursive: true });
  const sockPath = join(sockDir, 'mcp.sock');

  const chromeDir = join(home, '.claude', 'chrome');
  mkdirSync(chromeDir, { recursive: true });

  // Write chrome-native-host shim. Leave it writable so Claude Code's
  // --chrome flag can overwrite it with its own version (which will also
  // use the socat relay already running).
  const hostScript = join(chromeDir, 'chrome-native-host');
  try { chmodSync(hostScript, 0o755); } catch { /* doesn't exist yet */ }
  writeFileSync(hostScript, `#!/bin/sh\nexec node /usr/local/bin/chrome-mcp-bridge.cjs\n`);
  chmodSync(hostScript, 0o755);

  spawn('socat', [
    `UNIX-LISTEN:${sockPath},fork,reuseaddr`,
    `TCP:host.docker.internal:${bridgePort}`,
  ], { stdio: 'ignore', detached: true }).unref();

  log(`Chrome bridge: ${sockPath} → host.docker.internal:${bridgePort}`);

  // Write MCP server config so Claude Code picks up the chrome bridge on first start
  const settingsPath = join(home, '.claude', 'settings.json');
  let settings = {};
  if (existsSync(settingsPath)) {
    try {
      settings = JSON.parse(readFileSync(settingsPath, 'utf8'));
    } catch { /* start fresh */ }
  }
  settings.mcpServers = settings.mcpServers || {};
  settings.mcpServers['claude-in-chrome'] = {
    type: 'stdio',
    command: '/home/dev/.claude/chrome/chrome-native-host',
  };
  writeFileSync(settingsPath, JSON.stringify(settings, null, 2) + '\n');
  log('Wrote MCP config for claude-in-chrome');
}

// -- Auto-updater disable ----------------------------------------------------

function disableAutoUpdater() {
  const home = process.env.HOME || '/home/dev';
  const settingsPath = join(home, '.claude', 'settings.json');
  let settings = {};
  if (existsSync(settingsPath)) {
    try {
      settings = JSON.parse(readFileSync(settingsPath, 'utf8'));
    } catch {
      /* start fresh */
    }
  }
  settings.env = settings.env || {};
  settings.env.DISABLE_AUTOUPDATER = '1';
  writeFileSync(settingsPath, JSON.stringify(settings, null, 2) + '\n');
  log('Auto-updater disabled (updates via --rebuild)');
}

// -- Main --------------------------------------------------------------------

// Claude Code is installed at build time to /usr/local/bin/claude
if (which('claude')) {
  try {
    const ver = execFileSync('claude', ['--version'], { encoding: 'utf8' }).trim();
    log(`Claude Code: ${ver}`);
  } catch {
    log('Claude Code found but version check failed');
  }
} else {
  log('WARNING: Claude Code not found — rebuild the image with "devrig start --rebuild"');
}

if (process.env.BRIDGE_ENABLED === '1') {
  setupChromeBridge();
}
disableAutoUpdater();

// -- Devrig server (dashboard + API proxy) ------------------------------------

function startDevrigServer() {
  const script = '/usr/local/bin/devrig-server.js';
  if (!existsSync(script)) {
    log('WARNING: devrig-server.js not found — dashboard will not be available');
    return;
  }

  const logDir = join(process.env.HOME || '/home/dev', '.claude', 'logs');
  const logPath = join(logDir, 'devrig-server.log');
  mkdirSync(logDir, { recursive: true });

  // Respawn wrapper: restart the server when it exits (supports hot reload)
  function spawnServer() {
    const fd = openSync(logPath, 'a');
    const child = spawn('node', [script], {
      stdio: ['ignore', fd, fd],
      env: process.env,
    });
    closeSync(fd);
    child.on('exit', (code) => {
      log(`devrig-server exited (code ${code}), respawning...`);
      setTimeout(spawnServer, 500);
    });
  }
  spawnServer();

  log(`Devrig server starting on port ${process.env.DEVRIG_PORT || 8083}`);
}

startDevrigServer();

const sentinel = join(process.env.HOME || '/home/dev', '.claude', 'logs', '.setup-ready');
writeFileSync(sentinel, `ready ${Date.now() / 1000}\n`);
log('Setup complete — sentinel written');

File: .devrig/container-setup.py
#!/usr/bin/env python3
"""Docker container entrypoint setup: claude-code install/update + Chrome bridge."""

import glob
import logging
import os
import shutil
import subprocess
import time
from pathlib import Path

logging.basicConfig(format="[setup] %(asctime)s %(message)s", datefmt="%H:%M:%S", level=logging.INFO)
log = logging.getLogger()

NPM_PKG = "@anthropic-ai/claude-code"
NATIVE_INSTALLER_URL = "https://claude.ai/install.sh"


def _claude_version() -> str:
    return subprocess.run(["claude", "--version"], capture_output=True, text=True).stdout.strip()


def install_claude_code_npm():
    """Install/update Claude Code via npm."""
    prefix = os.environ["NPM_CONFIG_PREFIX"]
    pkg_dir = Path(prefix, "lib/node_modules/@anthropic-ai/claude-code")
    temp_glob = str(Path(prefix, "lib/node_modules/@anthropic-ai/.claude-code-*"))

    def cleanup_stale():
        if pkg_dir.exists():
            log.info("Removing stale install: %s", pkg_dir)
            shutil.rmtree(pkg_dir)
        for d in glob.glob(temp_glob):
            log.info("Removing temp dir: %s", d)
            shutil.rmtree(d)

    if shutil.which("claude"):
        log.info("claude found: %s", _claude_version())
        subprocess.run(["npm", "update", "-g", NPM_PKG], capture_output=True)
        log.info("claude after update: %s", _claude_version())
        return

    log.info("claude not found — installing %s", NPM_PKG)
    cleanup_stale()
    try:
        subprocess.run(["npm", "install", "-g", NPM_PKG], check=True)
    except subprocess.CalledProcessError:
        log.warning("First install attempt failed — cleaning cache and retrying")
        subprocess.run(["npm", "cache", "clean", "--force"], capture_output=True)
        cleanup_stale()
        subprocess.run(["npm", "install", "-g", NPM_PKG], check=True)

    log.info("Installed: %s", _claude_version())


def install_claude_code_native():
    """Install/update Claude Code via native installer."""
    if shutil.which("claude"):
        log.info("claude found: %s", _claude_version())
        log.info("Checking for updates...")
        subprocess.run(["claude", "update"], capture_output=True)
        log.info("claude after update: %s", _claude_version())
        return

    log.info("claude not found — installing via native installer")
    subprocess.run(
        ["bash", "-c", f"curl -fsSL {NATIVE_INSTALLER_URL} | bash"],
        check=True,
    )
    log.info("Installed: %s", _claude_version())


def install_claude_code():
    method = os.environ.get("CLAUDE_INSTALL_METHOD", "npm")
    if method == "native":
        install_claude_code_native()
    else:
        install_claude_code_npm()


def setup_chrome_bridge():
    bridge_port = os.environ.get("BRIDGE_PORT", "9229")
    user = os.environ["USER"]
    home = os.environ["HOME"]

    sock_dir = Path(f"/tmp/claude-mcp-browser-bridge-{user}")
    sock_dir.mkdir(parents=True, exist_ok=True)
    sock_path = sock_dir / "mcp.sock"

    chrome_dir = Path(home, ".claude/chrome")
    chrome_dir.mkdir(parents=True, exist_ok=True)

    host_script = chrome_dir / "chrome-native-host"
    host_script.write_text(
        f'#!/bin/bash\nexec node -e "process.stdin.pipe(require(\'net\').connect(\'{sock_path}\')).pipe(process.stdout)"\n'
    )
    host_script.chmod(0o755)

    subprocess.Popen(
        ["socat", f"UNIX-LISTEN:{sock_path},fork,reuseaddr", f"TCP:host.docker.internal:{bridge_port}"],
        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
    )
    log.info("Chrome bridge: %s → host.docker.internal:%s", sock_path, bridge_port)


if __name__ == "__main__":
    install_claude_code()
    setup_chrome_bridge()

    # Signal the host launcher that setup is complete.
    sentinel = Path(os.environ.get("HOME", "/home/dev")) / ".claude" / "logs" / ".setup-ready"
    sentinel.write_text(f"ready {time.time()}\n")
    log.info("Setup complete — sentinel written")
